### PR TITLE
internal(curves): add curve based on delta

### PR DESCRIPTION
**Pull request by elais**
This change creates a curve based on the delta (difference) between the
minimum and maximum values of a collection of curves.

As an example, basing fan speed on, say, water temperature - ambient
temperature will auto adjust the fan curve throughout the day as the
room heats up and cools down, keeping noise levels consistent and
eliminating a use case for fan profiles. It might not be useful for most
users but it is a nice option to have.

### CODE REVIEW

**Comment by markusressel** [comment]
In its current implementation the diff is always >= 0, not sure if this might create problems. In your example it probably wont, since the air cannot get cooler than ambient, but for other use cases it might?

**Comment by elais** [comment]
I did think about that, and considered taking the abs of the diff so `delta := math.Abs(dmax-dmin)` which would just invert the numbers if we ever get a diff <= 0.

What do you think?

**Comment by elais** [comment]
Thinking more about this, the one use case I can think of where air can get cooler than ambient is with the use of a chiller. In that case there likely isn't a problem here because pwm signals can only range from 0-255. The case where pwm < 0 doesn't exist.

**Comment by elais** [comment]
I looked into that test run error and it seems to only happen when the tests are run concurrently. Is there a global being shared between the tests that gets overwritten?

**Comment by markusressel** [comment]
Just to make sure I get this right: 

Your desired curve should find both the biggest and the smallest value, and then print the difference between them (which cannot be negative). 

This, however, doesn't always make sense, since the meaning of `60, 90` is the complete opposite of `90, 60`, but both would return the same diff of 30 (and therefore result in the same curve value -> fan speed).

Idea: The first sensor always functions as the baseline sensor/value, which every other sensor/value is diffed against. All diffs are sorted increasing and the biggest diff (no matter if positive or negative) wins/is returned.

Also: keep in mind that the function curve only accepts other **curve** ids, and that curves _always_ map sensor values to a range of [0..255] (which is then again mapped to the valid PWM range of the associated fan). So, currently you cannot really do:

```
curves:
  - id: diff
    function:
      type: diff
      curves:
        - ambient_temp_sensor
        - cpu_sensor
```

You would have to create intermediate curves instead, and also create a range for them:

```
curves:
  - id: diff
    function:
      type: diff
      curves:
        - ambient_curve
        - cpu_curve

  - id: cpu_curve
    linear:
      sensor: cpu_sensor
      min: 40
      max: 80

  - id: ambient_curve
    linear:
      sensor: ambient_temp_sensor
      min: 16
      max 30
```

**Comment by elais** [comment]
This is how I understand this working based on the test case


```yaml

curves:
    - id: water_temp_curve
        linear:
          sensor: water_temp_sensor
          min: 18
          max: 60

  - id: ambient_curve
      linear:
        sensor: ambient_temp_sensor
        min: 18
        max: 60
```

the curves are created using the function `createLinearCurveConfig`

```go
	curve1 :=  createLinearCurveConfig(
		"case_fan_front",
		s1.GetId(), // s1 being the water_temp_sensor
		18,
		60,
	)

	curve2 := createLinearCurveConfig(
		"case_fan_back",
		s2.GetId(), // s2 being the water_temp_sensor
		18,
		60,
	)
```
The two curves are fed into the function `createFunctionCurveConfig`, which creates an object that includes the function `Evaluate`. `Evaluate` takes the current temperature and evaluates a pwm value based on the curve and the temperature sensor reading for each curve and creates an array of values representing each curve's pwm value at the current temperature. Then we take that array of values and apply our function to the array. So if a result is a two element array with the values 20, and 150, the function will of course do 150 - 20.

In the test case that I wrote, we took as input two linear curves based on two separate sensors and applied our function under the assumption that we would subtract the smallest pwm reading from the highest pwm reading in the collection.

> This, however, doesn't always make sense, since the meaning of 60, 90 is the complete opposite of 90, 60, but both would return the same diff of 30 (and therefore result in the same curve value -> fan speed).

This was intentional. If we're basing fan speed on the difference between two curves, then it doesn't really matter which curve is higher. If the difference is small, the resulting fan speeds will be lower, if the difference is large, the fan speeds will be higher. We only care about the distance between the numbers, not really the values themselves.


> And even then it isn't complete, since one would have to provide a range of "diff values" to map to the [0..255] range.

Why is this? According to the test cases the function is applied to pwm values that are derived from the current temperature being read on the sensors. Is this not the case?

> Also: keep in mind that the function curve only accepts other curve ids, and that curves always map sensor values to a range of [0..255] (which is then again mapped to the valid PWM range of the associated fan).

The curves map sensor values to a range but my reading of the code says the function is applied only to the collection of curve values that are calculated at the current temperature. Could you look at my test case in `curve_test.go`. When I change my input temperatures it also changes the pwm value I receive from `functionCurve.Evaluate()`. Am I not looking at this correctly?

**Comment by markusressel** [comment]
> Why is this? According to the test cases the function is applied to pwm values that are derived from the current temperature being read on the sensors. Is this not the case?

Sorry, when I wrote this I was thinking about implementing a separate curve type, next to the function one (which could accept sensors instead of curves). You are right, the function is applied to the values of the other curves, which already are in [0..255], so there is no additional mapping needed. Sorry for the confusion.

> This was intentional. If we're basing fan speed on the difference between two curves, then it doesn't really matter which curve is higher. If the difference is small, the resulting fan speeds will be lower, if the difference is large, the fan speeds will be higher. We only care about the distance between the numbers, not really the values themselves.

The curve is not limited to two curves though, it would also be possible to compare three curves. I just have the feeling that this limits the possible use cases for this curve type :thinking: But maybe I am overthinking this and we should just try and see what can and can't be done with it.

File: internal/curves/curve.go [hunk 1]
```diff
@@ -106,6 +106,16 @@ func (c functionSpeedCurve) Evaluate() (value int, err error) {
...
 	switch c.function {
+    case configuration.FunctionDelta:
+		var dmax = float64(values[0])
+		var dmin = float64(values[0])
+		for _, v := range values {
+			var val = float64(v);
+			dmin = math.Min(dmin, float64(val))
+			dmax = math.Max(dmax, float64(val))
+		}
+		delta := math.Abs(dmax - dmin)
```
Comment (markusressel) [inline internal/curves/curve.go:117]: This `math.Abs` doesn't add anything. You always substract `dmin` from `dmax`, where `dmin !<= dmax`. So the "worst case" is `delta == 0` and `delta == 255` for the opposite direction. 
That was how I initially got my concern from :smile: 
Or am I missing something?

File: internal/curves/curve.go [hunk 2]
```diff
@@ -106,6 +106,16 @@ func (c functionSpeedCurve) Evaluate() (value int, err error) {
 	}
 
 	switch c.function {
+    case configuration.FunctionDelta:
+		var dmax = float64(values[0])
+		var dmin = float64(values[0])
+		for _, v := range values {
+			var val = float64(v);
```
Comment (markusressel) [inline internal/curves/curve.go:113]: The semicolon is unnecessary,

and please avoid this temporary variable and just use `float64(v)` down below.

File: internal/curves/curve_test.go [hunk 3]
```diff
@@ -205,6 +205,65 @@ func TestFunctionCurveAverage(t *testing.T) {
...
+		Name:      "sensor_water",
+		MovingAvg: temp2,
+	}
+	sensors.SensorMap[s2.GetId()] = &s2
+
+	curve1 :=  createLinearCurveConfig(
+		"case_fan_front",
+		s1.GetId(),
+		18,
+		60,
```
Comment (markusressel) [inline internal/curves/curve_test.go:231]: I don't really understand how you choose those values, but it does work as expected so... :+1:  :sweat_smile:

File: internal/configuration/curves.go [hunk 4]
```diff
@@ -15,6 +15,7 @@ type LinearCurveConfig struct {
 
 const (
 	FunctionAverage = "average"
+	FunctionDelta   = "delta"
```
Comment (markusressel) [inline internal/configuration/curves.go:18]: We have to add this new possibile function type to the README and example config file documentation too.

**Review by markusressel — CHANGES_REQUESTED** [review]

**Comment by elais** [comment]
> The curve is not limited to two curves though, it would also be possible to compare three curves. I just have the feeling that this limits the possible use cases for this curve type 🤔 But maybe I am overthinking this and we should just try and see what can and can't be done with it.

Honestly I just see this as an implementation of range, a standard aggregate function. Given the curve functions all operate on aggregate data we could cover most use cases by just implementing as many  aggregate functions as possible and allowing users to pipeline them into weird custom curves. I just did this one because (selfishly) it fits my use case. 

Also about the unit tests. Could you take a look at curve_test.go and see what's causing the text failures on concurrent runs? If I force tests to run sequentially the problem goes away.

**Comment by markusressel** [comment]
Yeah the global storage for sensors, curves and fans is probably not helping here because multiple tests use the same name (id) for curves. So when run concurrently, they probably override each other. Fix would be to use a unique name for the test curve in your test case (`case_fan_front_123` and `case_fan_back_123`). Not sure if I can get rid of this global map at some point, currently it is necessary.

File: internal/curves/curve_test.go (hunk 3 above)
Comment (elais) [reply internal/curves/curve_test.go:231]: I don't remember, I think I as testing different numbers and just checked those in, lol.

**Comment by elais** [comment]
added the requested changes. Is it ok if we run the tests sequentially for now? Changing the case fan names results in panic

**Comment by markusressel** [comment]
I checked on my machine just a few minutes ago. Can I push to your branch? I will fix it then.

**Comment by elais** [comment]
You'd have to open a pull request to push to mine. It may be easier to push to your master and bypass the CI for now since we've both run the tests and they pass. That's enough verification for me but ymmv.

Besides I've already shut down for the night.

**Comment by markusressel** [comment]
I was able to push to your branch.
Thx for your contribution! :heart:

**Comment by elais** [comment]
Thank you!

**Pull request merged**